pkg/rpcclient/nep24/royalty.go

```diff
@@ -0,0 +1,182 @@
+// Package nep24 provides RPC wrappers for NEP-24 contracts, which define NFT royalties.
+package nep24
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/nspcc-dev/neo-go/pkg/neorpc/result"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient/nep11"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient/neptoken"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
+)
+
+// RoyaltyRecipient contains information about the recipient and the royalty amount.
+type RoyaltyRecipient struct {
+	Address util.Uint160
+	Amount  *big.Int
+}
+
+// RoyaltiesTransferredEvent represents a RoyaltiesTransferred event as defined in the NEP-24 standard.
+type RoyaltiesTransferredEvent struct {
+	RoyaltyToken     util.Uint160
+	RoyaltyRecipient util.Uint160
+	Buyer            util.Uint160
+	TokenID          []byte
+	Amount           *big.Int
+}
+
+// RoyaltyReader is an interface for contracts implementing NEP-24 royalties.
+type RoyaltyReader struct {
+	neptoken.Base
+	Invoker nep11.Invoker
+	hash    util.Uint160
+}
+
+// Royalty is a full reader interface for NEP-24 royalties.
+type Royalty struct {
+	RoyaltyReader
+}
+
+// NewRoyaltyReader creates an instance of RoyaltyReader for a contract with the given hash using the given invoker.
+func NewRoyaltyReader(invoker nep11.Invoker, hash util.Uint160) *RoyaltyReader {
+	return &RoyaltyReader{
+		Base:    *neptoken.New(invoker, hash),
+		Invoker: invoker,
+		hash:    hash,
+	}
+}
+
+// NewRoyalty creates an instance of Royalty for a contract with the given hash using the given actor.
+func NewRoyalty(actor nep11.Actor, hash util.Uint160) *Royalty {
+	return &Royalty{
+		RoyaltyReader: *NewRoyaltyReader(actor, hash),
+	}
+}
+
+func (c *RoyaltyReader) RoyaltyInfo(tokenID []byte, royaltyToken util.Uint160, salePrice *big.Int) ([]RoyaltyRecipient, error) {
+	res, err := c.Invoker.Call(c.hash, "royaltyInfo", tokenID, royaltyToken, salePrice)
+	if err != nil {
+		return nil, err
+	}
+	var royalties []RoyaltyRecipient
+	for _, item := range res.Stack {
+		royalty, ok := item.Value().([]stackitem.Item)
+		if !ok || len(royalty) != 2 {
+			return nil, fmt.Errorf("invalid royalty structure: expected array of 2 items, got %d", len(royalty))
+		}
+		var recipient RoyaltyRecipient
+		err = recipient.FromStackItem(royalty)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode royalty detail: %w", err)
+		}
+		royalties = append(royalties, recipient)
+	}
```

Comment:
Is it the auto-generated code or hand-written? This code is not correct, because contract returns a single stackitem (array of RoyaltyRecipient).

Reply:
Waiting for the response.

Reply:
it is hand-written, but as was discussed it should not be like that.

Reply:
Fix it, the resulting stack must contain exactly a single item, this item is an Array containing Structs (or Arrays). These structs are RoyaltiesInfo.

The current version of this code waits two items on stack.